.vsts.pipelines/jobs/prepare-artifacts.yml

```diff
@@ -0,0 +1,84 @@
+parameters:
+  gatherJobs: []
+  gatherPortableJob: ''
+
+  # downloadBuildConfig by default uses the current build, but can be changed to point at an
+  # existing/previous build by passing properties like these, instead:
+  #
+  #   buildType: specific
+  #   buildVersionToDownload: specific
+  #   project: 'public'
+  #   definition: 'source-build-CI'
+  #   buildId: 741400
+  #
+  # By defining this in one place, it's easier to set this up if necessary. This may be useful
+  # because the build takes a long time relative to prepare-artifacts, and we may want to tweak
+  # prepare-artifacts in case it fails due to a bug in the future. ...It's also nice to have the
+  # properties listed in this comment even if we don't use them here, because it's hard to figure
+  # them out from the docs alone--it'll be a nice reference.
+  downloadBuildConfig:
+    buildType: current
+
+jobs:
+- job: PrepareArtifacts
+  pool:
+    vmImage: 'ubuntu-18.04'
+  dependsOn:
+  - ${{ each gather in parameters.gatherJobs }}:
+    - ${{ gather.job }}
+  timeoutInMinutes: 300
+  workspace:
+    clean: all
+  variables:
+    artifactStageDir: '$(Build.SourcesDirectory)/artifacts/stage'
+    nonportableSourceBuiltStageDir: '$(artifactStageDir)/nonportableSourceBuilt'
+    portableSourceBuiltStageDir: '$(artifactStageDir)/portableSourceBuilt'
+    allSourceBuiltStageDir: '$(artifactStageDir)/allSourceBuilt'
+    outputTarballFile: '$(artifactStageDir)/Private.SourceBuilt.Artifacts.$(Build.BuildId).tar.gz'
+  steps:
+  - ${{ each gather in parameters.gatherJobs }}:
+    - task: DownloadBuildArtifacts@0
+      displayName: 'Download ${{ gather.job }} nonportable'
+      inputs:
+        ${{ insert }}: ${{ parameters.downloadBuildConfig }}
+        downloadType: single
+        artifactName: ${{ coalesce(gather.artifactName, format('Tarball {0} Offline', gather.job)) }}
+        downloadPath: $(nonportableSourceBuiltStageDir)
+        allowPartiallySucceededBuilds: true
+
+  - task: DownloadBuildArtifacts@0
+    displayName: 'Download ${{ parameters.gatherPortableJob }} portable'
+    inputs:
+      ${{ insert }}: ${{ parameters.downloadBuildConfig }}
+      downloadType: single
+      artifactName: 'Tarball ${{ parameters.gatherPortableJob }} Offline Portable'
+      downloadPath: $(portableSourceBuiltStageDir)
+      allowPartiallySucceededBuilds: true
+
+  - script: |
+      find "$(artifactStageDir)" -type f -exec du -h {} \;
+    displayName: Show downloaded artifacts
+
+  - script: |
+      set -xeuo pipefail
+      mkdir -p "$(allSourceBuiltStageDir)"
+
+      # Extract all source-built assets into a single place. Overlap and overwrites are expected.
+      # What matters is the portable ones are copied last and ultimately win.
+      find \
+        "$(nonportableSourceBuiltStageDir)" \
+        "$(portableSourceBuiltStageDir)" \
+        -iname 'Private.SourceBuilt.Artifacts.*.tar.gz' \
+        -exec tar -xf {} -C "$(allSourceBuiltStageDir)" \;
```

Comment:
What about PackageVersions.props file?  If last one in wins, it will not reflect the combined files.  It seems like this needs to be coupled with https://github.com/dotnet/source-build/issues/1638

Reply:
This is how @crummel made the previously-source-built tarball we're using now, and I don't see any PVPs, so I assume it worked out... I'm not sure.

Reply:
Yeah, it's not strictly correct, but it works out because we don't use runtime package versions for anything right now.  If we're productizing it, it might be worth looking into a bit.  There's probably two things we care about:
- getting all package versions into PackageVersions.props: I think something like adding unique lines from each individual PackageVersions.props to the output PackageVersions.props would work decently well here and probably shouldn't screw up too much.  This currently doesn't matter but would cover us if we ever needed to depend on, e.g., a runtime package's version.
- not having any conflicting package versions: in theory, this shouldn't happen.  Package names and versions should be consistent across RIDs, and I don't think we've seen any real-life cases where this has gone wrong.  I can't think offhand of a cheap way to detect this other than a bunch of `sed` nonsense, but I think it would be worth filing an issue for a future improvement.

I don't think either of these are blockers and #1638 should obviate them anyway.